Initial version of visualization tool

sbmacdonald opened the pull request:
## Sample interactive python usage:
![image](https://cloud.githubusercontent.com/assets/2412361/22550189/55870676-e925-11e6-8c8e-758be9d3347b.png)


runs with `view=True` results in (on my machine) Gwenview opening wtih the image displayed:

![image](https://cloud.githubusercontent.com/assets/2412361/22342972/aa717968-e3cc-11e6-9268-d896b7d6bbab.png)

Nodes identified with an abnormal value for a term will be highlighted by using a fill color== red:

 ![743ec297-fc9d-216f-d9b3-dfc7d9900d9c_lr_with_terms](https://cloud.githubusercontent.com/assets/2412361/22600937/26cfd68e-ea13-11e6-9560-a65f1208dee3.png)

Additional samples:

## 1. `render(rankdir='TB')`
![14881da6-ee85-4674-96a6-8ec0d35cb176_tb_with_terms](https://cloud.githubusercontent.com/assets/2412361/22662842/d0f1f740-ec80-11e6-8429-78515ce0cd18.png)

## 2. `render(rankdir='LR', compact=False)`
![14881da6-ee85-4674-96a6-8ec0d35cb176_lr_not_compact_with_terms](https://cloud.githubusercontent.com/assets/2412361/22662843/d0f21f22-ec80-11e6-9e52-982a86fe28dc.png)

## 3. `render(rankdir='LR', with_terms=False)`
![14881da6-ee85-4674-96a6-8ec0d35cb176_lr](https://cloud.githubusercontent.com/assets/2412361/22662846/d0f5fe8a-ec80-11e6-8c53-98cd6f4bd4d2.png)

## 4.`render(rankdir='LR', warnings=False)`
![14881da6-ee85-4674-96a6-8ec0d35cb176_lr_with_terms](https://cloud.githubusercontent.com/assets/2412361/22662844/d0f2f064-ec80-11e6-9894-b94002bce57f.png)

sbmacdonald committed:
Initial version of visualization tool

sbmacdonald committed:
refactoring to improve oo-ness

sbmacdonald committed:
Incremental improvements

- Added some tests
- simplified _format_<> helpers
- improved node formatting

sbmacdonald committed:
Added verbose option

sbmacdonald committed:
Additional tests, and bug fix

sbmacdonald committed:
Travis could not find graphviz

sbmacdonald committed:
Merge branch 'master' into XX-initial-version

sbmacdonald committed:
Tweak filename format

sbmacdonald committed:
Added verbose option to fromId()

sbmacdonald committed:
Added participation, reinstatements and better formatting

bmacphee commented on analyzere_extras/visualizations.py:
```diff
@@ -0,0 +1,341 @@
+            # hash the current node to see if it is unique
+            node_hash = hashlib.md5(str(l).encode('utf-8')).hexdigest()
+            if node_hash not in unique_nodes:
+                unique_nodes[node_hash] = next(sequence)
```
consider changing how sequence is implemented:
```
>>> a = itertools.count()
>>> next(a)
0
>>> next(a)
1
>>> next(a)
2
```
you can do `itertools.count(start=1)` if you need that

sbmacdonald committed:
Improvements

- Improved vebose performance
- Added 'warning' from _format_layer_terms so we can render
"warning" nodes in red

bmacphee commented on analyzere_extras/visualizations.py:
```diff
@@ -0,0 +1,341 @@
+        else:
+            terms += '-inf'
+
+        if hasattr(layer, 'expiry_date'):
```
for OCD compatible formatting, might want to put the attr outputs into a list then
`', '.join()` the list

bmacphee replied:
it could kill the extra if/else nesting by doing so, since you could (at the end) check `if date_list: terms += coverage + joined:`

bmacphee replied:
apply any of these techniques below where appropriate...

bmacphee commented on tests/test_visualizations.py~:
```diff
@@ -0,0 +1,11 @@
+import pytest
```
accidentally committed file?

bmacphee commented on tests/test_visualizations.py:
```diff
@@ -0,0 +1,308 @@
+        assert visualizations._format_description(raw) == expected
+
+    def test_description_with_windows_file_path(self):
+        raw = 'Layer 2 : Z:\\losses\\alpha.csv'
```
```
>>> r'Z:\\losses\\alpha.csv'
'Z:\\\\losses\\\\alpha.csv'
```
Using r'' for your string delimiters can make things easier on the eyes with escapes (r mnemonic == raw string; i.e. no escape interpolation)

bmacphee commented on tests/test_visualizations.py:
```diff
@@ -0,0 +1,308 @@
+            lvg = visualizations.LayerViewDigraph(lv)
+
+    def test_basic(self, layer_view):
+        lvg = visualizations.LayerViewDigraph(layer_view)
```
unused local.  could just 
`assert visualizations.LayerViewDigraph(layer_view)` if you expect it to return not None

bmacphee commented on tests/test_visualizations.py:
```diff
@@ -0,0 +1,308 @@
+        m = {'something': 'here'}
+        lv = convert_to_analyzere_object(m, Resource)
+        with pytest.raises(ValueError):
+            lvg = visualizations.LayerViewDigraph(lv)
```
unused local.  could just kill it.  sometimes if you are forced to make an assignment like in a loop counter or whatever, the python conventions tend to use `_` as a variable name meaning 'variable I don't care about'

bmacphee commented on tests/test_visualizations.py:
```diff
@@ -0,0 +1,308 @@
+        lvg = visualizations.LayerViewDigraph(layer_view, format='svg')
+        assert lvg._graph.format == 'svg'
+
+    def test_render(self, layer_view):
```
Not sure how much pain this would involve, but to avoid having to write to the file system in tests and clean up, using `Mock()` for the `_graph` member and just checking the arguments to `_graph.render()` or whatever that the correct filename is used should be enough... maybe?

Assuming we can trust the graphviz library to do what it promises... but we should be able to.

bmacphee reviewed (changes requested):
I think at least some of my suggestions are important enough for me to officially request changes.  So that's what I selected for the review option.

sbmacdonald committed:
Flake8
More tests

sbmacdonald committed:
Code review changes

- Test refactoring with Mock
- Additional tests
- add tests to flak8

bmacphee commented on analyzere_extras/visualizations.py:
```diff
@@ -0,0 +1,352 @@
+
+
+def _format_description(description):
+    """Clean the description of a node for diaply in Graphviz"""
```
`diaply` => `display`

bmacphee commented on analyzere_extras/visualizations.py:
```diff
@@ -0,0 +1,352 @@
+
+def _format_MoneyField(mf):
+    """Format a MoneyField to '<unlimited|value> CCY' """
+    formatted = 'n/a'
```
this line not required in python

sbmacdonald committed:
Added additional tests for Terms formatting

sbmacdonald committed:
removed extraneous assignment

sbmacdonald committed:
Added Usage section

bmacphee committed:
refactored text formatting

bmacphee committed:
hid warning state inside helper

sbchisholm commented on README.rst:
```diff
@@ -1,23 +1,61 @@
+
+   from analyzere import LayerView
+
+   lv = analyzere.LayerView.retrieve('011785b1-203b-696e-424e-7da9b0ec779a')
```
Is there a way to add syntax highlighting to these code blocks?

sbmacdonald replied:
can't seem to find anything that works

DanielArndt replied:
I don't think it works on github (for whatever reason) but the [code and codeblocks directives](https://github.com/github/markup/issues/106) allow you to specificy a language.

If I remember correctly, this worked well in sphinx, but not on github.

sbmacdonald replied:
yeah tried it and github doesn't support it

sbchisholm replied:
no worries

sbchisholm commented on analyzere_extras/visualizations.py:
```diff
@@ -0,0 +1,371 @@
+        self._generate_nodes(lv.layer, sequence, unique_nodes, edges)
+
+    @staticmethod
+    def fromId(lv_id, with_terms=True, verbose=False,
```
I think the typical python convection for this function name would be `from_id`, was there a reason you used camel case here?

bmacphee committed:
fully separated responsibility for term formatting

sbmacdonald committed:
Added 'participation'
Added 'franchise'
Moved 'nth'
Added tests that ensure order is maintained

sbmacdonald committed:
Added 'warnings' option to allow users to disable warning nodes
being colored red

sbmacdonald committed:
Merge branch 'XX-initial-version' of github.com:analyzere/analyzere-python-extras into XX-initial-version

sbchisholm commented on analyzere_extras/visualizations.py:
```diff
@@ -0,0 +1,381 @@
+                        from bottom to top, and from right to left,
+                        respectively.
+        """
+        # This raise and exception if any of the following analyzere variables
```
wording: This will raise an exception....

sbchisholm commented on analyzere_extras/visualizations.py:
```diff
@@ -0,0 +1,381 @@
+
+        except RuntimeError:
+            # native display failed, revert to returning a clickable FileLink
+            # iff we arein a python notebook, otherwise return the filename.
```
sp: are in

sbchisholm reviewed (commented):
Look good, just a few comments.

sbmacdonald committed:
Review comments
changed verbose -> compact
fixed not_compact rendering

DanielArndt commented on analyzere_extras/visualizations.py:
```diff
@@ -0,0 +1,383 @@
+        self._terms = OrderedDict()
+        self.warning = False
+
+    def append(self, required_attr, attr_formatted=None,
```
@bmacphee I'd rather see this (`attr_formatted`) called something like `display_name` personally, but that's preference.

sbmacdonald committed:
s/attr_formatted/display_name/g
